File: Q0437_Path_Sum_III/Solution.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        if root is None:
            return 0
        res = self.find_path(root, sum)
        res += self.pathSum(root.left, sum)
        res += self.pathSum(root.right, sum)
        return res

    def find_path(self, root: TreeNode, sum: int) -> int:

        if root is None:
            return 0
        res = 0
        if root.val == sum:
            res += 1
        res += self.find_path(root.left, sum - root.val)
        res += self.find_path(root.right, sum - root.val)
        return res

File: Q0437_Path_Sum_III/Solution.java
package Q0437_Path_Sum_III;

/**
 * @ Description:
 * @ Date: Created in 21:40 2019-04-14
 * @ Author: Anthony_Duan
 */
public class Solution {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    public int pathSum(TreeNode root, int sum) {

        if (root == null) {
            return 0;
        }
        int res = findPath(root, sum);
        res += pathSum(root.left, sum);
        res += pathSum(root.right, sum);
        return res;

    }

    private int findPath(TreeNode root, int sum) {
        if (root == null) {
            return 0;
        }
        int res = 0;
        //因为这里可能存在负值
        if (root.val == sum) {
            res += 1;
        }

        res += findPath(root.left, sum - root.val);
        res += findPath(root.right, sum - root.val);
        return res;
    }

}
